File: launchschool_exercises/109_small_problems/easy8/sum_of_sums.rb
def sum_of_sums(arr)
  new_arr = []
  arr.each_index do |x|
    next_element = []
    (0..x).each { |y| next_element << arr[y] }
    new_arr << next_element
  end
  new_arr.flatten.sum
end

p sum_of_sums([3, 5, 2]) == (3) + (3 + 5) + (3 + 5 + 2) # -> (21)
p sum_of_sums([1, 5, 7, 3]) == (1) + (1 + 5) + (1 + 5 + 7) + (1 + 5 + 7 + 3) # -> (36)
p sum_of_sums([4]) == 4
p sum_of_sums([1, 2, 3, 4, 5]) == 35

File: launchschool_exercises/210_small_problems/list_processing/sum-of-sums.js
/*
 Write a function that takes an array of numbers, and returns the sum of the sums of each leading subsequence for that array. You may assume that the array always contains at least one number.
*/

function sumOfSums(arr) {
  total = 0;

  for (let i = 0; i < arr.length; i++) {
    total += arr.slice(0, i + 1).reduce((acc, curr) => acc += curr);
  }

  return total;
}

console.log(sumOfSums([3, 5, 2]));        // (3) + (3 + 5) + (3 + 5 + 2) --> 21
console.log(sumOfSums([1, 5, 7, 3]));     // (1) + (1 + 5) + (1 + 5 + 7) + (1 + 5 + 7 + 3) --> 36
console.log(sumOfSums([4]));              // 4
console.log(sumOfSums([1, 2, 3, 4, 5]));  // 35
